### ISSUE TITLE
Some questions about code in stack_llama/supervised_finetuning.py

### ISSUE THREAD

**Issue by Kororinpas** [lvwerra/trl#291]
`class ConstantLengthDataset(IterableDataset):
    """
    Iterable dataset that returns constant length chunks of tokens from stream of text files.
        Args:
            tokenizer (Tokenizer): The processor used for proccessing the data.
            dataset (dataset.Dataset): Dataset with text files.
            infinite (bool): If True the iterator is reset after dataset reaches end else stops.
            seq_length (int): Length of token sequences to return.
            num_of_sequences (int): Number of token sequences to keep in buffer.
            chars_per_token (int): Number of characters per token used to estimate number of tokens in text buffer.
    """

    def __init__(
        self,
        tokenizer,
        dataset,
        infinite=False,
        seq_length=1024,
        num_of_sequences=1024,
        chars_per_token=3.6,
    ):
        self.tokenizer = tokenizer
        self.concat_token_id = tokenizer.eos_token_id if tokenizer.eos_token_id else args.eos_token_id
        self.dataset = dataset
        self.seq_length = seq_length
        self.infinite = infinite
        self.current_size = 0
        self.max_buffer_size = seq_length * chars_per_token * num_of_sequences

    def __iter__(self):
        iterator = iter(self.dataset)
        more_examples = True
        while more_examples:
            buffer, buffer_len = [], 0
            while True:
                if buffer_len >= self.max_buffer_size:
                    break
                try:
                    buffer.append(prepare_sample_text(next(iterator)))
                    buffer_len += len(buffer[-1])
                except StopIteration:
                    if self.infinite:
                        iterator = iter(self.dataset)
                    else:
                        more_examples = False
                        break
            tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
            all_token_ids = []
            for tokenized_input in tokenized_inputs:
                all_token_ids.extend(tokenized_input + [self.concat_token_id])
            for i in range(0, len(all_token_ids), self.seq_length):
                input_ids = all_token_ids[i : i + self.seq_length]
                if len(input_ids) == self.seq_length:
                    self.current_size += 1
                    yield {
                        "input_ids": torch.LongTensor(input_ids),
                        "labels": torch.LongTensor(input_ids),
                    }`

I would like to ask a question about the last part of the code. As far as I understand, the loop at the end of the code is truncating the text based on seq_length, without considering the question-answer format in the dataset. This means that the truncated text may contain incomplete question-answer pairs. I would like to inquire whether this truncation is intentional and whether it could potentially decrease the effectiveness of the training compared to using complete question-answer pairs as input_ids.

**Comment by lvwerra** [comment]
This is a bit a trade-off between training efficiently and potentially corrupting a few examples: if you are limited by the number of examples rather than compute then you might want to avoid this and just have one text per sample in the batch and pad the batch. However, if you have a lot of data the above code is very efficient since there is no padding at all. True, some examples might not have the full context but usually in the long training this doesn't matter much. The strategy above is called packing and common practice when pretraining language models. Hope this helps!

**Comment by Kororinpas** [comment]
@lvwerra Thank you very much for your patient response. I also saw your explanation of the packing technique in your post https://huggingface.co/blog/stackllama, and combined with your reply, I now have a deeper understanding of the packing technique. Thank you again
